### PR TITLE
*: add explicit partition pruning to index joins (#32007)

### LINKED ISSUE #32007: partition as subquery with `<> all` got wrong result under dynamic pruning mode
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
set tidb_partition_prune_mode='dynamic';
drop table if exists t1, t2, t3;
create table t1 (a int, b tinyint, primary key (a)) partition by range (a) (
    partition p0 values less than (5),
    partition p1 values less than (20),
    partition p2 values less than (30),
    partition p3 values less than (40),
    partition p4 values less than MAXVALUE
);
insert into t1 values (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (6, 6), (7, 7), (10, 10), (11, 11), (12, 12), (13, 13), (14, 14), (15, 15), (20, 20), (21, 21), (22, 22), (23, 23), (24, 24), (25, 25), (30, 30), (31, 31), (32, 32), (33, 33), (34, 34), (35, 35), (36, 36), (40, 40), (50, 50), (80, 80), (90, 90), (100, 100);
create table t3 (a int, b mediumint, primary key (a));
insert into t3 values (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (6, 6), (7, 7), (8, 8), (9, 9), (10, 10), (11, 11), (12, 12), (13, 13), (14, 14), (15, 15), (16, 16), (17, 17), (18, 18), (19, 19), (20, 20), (21, 21), (22, 22), (23, 23);

select * from t3 where t3.a <> ALL (select t1.a from t1 partition (p0)) order by t3.a;
```

### 2. What did you expect to see? (Required)
```sql
MySQL root@172.16.5.103:table_partition> select * from t3 where t3.a <> ALL (select t1.a from t1 partition (p0)) order by t3.a;
+----+----+
| a  | b  |
+----+----+
| 5  | 5  |
| 6  | 6  |
| 7  | 7  |
| 8  | 8  |
| 9  | 9  |
| 10 | 10 |
| 11 | 11 |
| 12 | 12 |
| 13 | 13 |
| 14 | 14 |
| 15 | 15 |
| 16 | 16 |
| 17 | 17 |
| 18 | 18 |
| 19 | 19 |
| 20 | 20 |
| 21 | 21 |
| 22 | 22 |
| 23 | 23 |
+----+----+
```
### 3. What did you see instead (Required)
```sql
MySQL root@127.0.0.1:test> select * from t3 where t3.a <> ALL (select t1.a from t1 partition (p0)) order by t3.a;
+----+----+
| a  | b  |
+----+----+
| 8  | 8  |
| 9  | 9  |
| 16 | 16 |
| 17 | 17 |
| 18 | 18 |
| 19 | 19 |
+----+----+

```
### 4. What is your TiDB version? (Required)
```sql
MySQL root@127.0.0.1:test> select tidb_version()\G
***************************[ 1. row ]***************************
tidb_version() | Release Version: v5.5.0-alpha-234-g19a020c59-dirty
Edition: Community
Git Commit Hash: 19a020c59956d1f430c41ca9d168a7417dc33490
Git Branch: master
UTC Build Time: 2022-01-27 03:45:36
GoVersion: go1.17.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #32007 

Problem Summary:
Explicit partition selection `SELECT ... FROM pt PARTITION(p) ...` was not considered when buildTableReaderForIndexJoin().

### What is changed and how it works?
Use the PartitionNames (if set) and do skip looking up contents in non-mentioned partitions.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fixed explicit partition selection in subselects.
```

### CODE REVIEW

**Review by djshow832 — COMMENTED** [review]
I'm curious: does only `IndexJoin` have this problem?

**Comment by mjonss** [comment]
> I'm curious: does only `IndexJoin` have this problem?

Great question! Not sure, but I just tried with Hash join and Merge join for the same query, and it does give the correct result.
```
select /*+ HASH_JOIN(t3) */ * from t3 where t3.a <> ALL (select t1.a from t1 partition (p0)) order by t3.a;
select /*+ MERGE_JOIN(t3) */ * from t3 where t3.a <> ALL (select t1.a from t1 partition (p0)) order by t3.a;
```
Notice that the `EXPLAIN` of the original query says it only uses `partition:p0` which is wrong (i.e. the explain code takes uses the given `partition (p0)` but the index join does not implement it.

@djshow832 Which other types of joins needs to be tested?

File: executor/builder.go [hunk 1]
```diff
@@ -3941,6 +3964,9 @@ func (builder *dataReaderBuilder) buildTableReaderForIndexJoin(ctx context.Conte
 		}
 		for _, p := range partitions {
 			pid := p.GetPhysicalID()
+			if _, ok := explicitPartitionSelection[pid]; !ok {
+				continue
+			}
```
Comment (djshow832) [inline executor/builder.go:3969]: I think the partitions are already pruned by `partitionProcessor.convertToIntSlice()` indirectly through `partitionPruning()`.
Thus, I think the code branches where `partitionPruning()` is not called may have the problem.

**Comment by djshow832** [comment]
> > I'm curious: does only `IndexJoin` have this problem?
> 
> Great question! Not sure, but I just tried with Hash join and Merge join for the same query, and it does give the correct result.
> 
> ```
> select /*+ HASH_JOIN(t3) */ * from t3 where t3.a <> ALL (select t1.a from t1 partition (p0)) order by t3.a;
> select /*+ MERGE_JOIN(t3) */ * from t3 where t3.a <> ALL (select t1.a from t1 partition (p0)) order by t3.a;
> ```
> 
> Notice that the `EXPLAIN` of the original query says it only uses `partition:p0` which is wrong (i.e. the explain code uses the given `partition (p0)` but the index join does not implement it.
> 
> @djshow832 Which other types of joins needs to be tested?

Not only joins but also other operators, which are hard to be tested fully. So I think the better way is to go through the code to check why `buildTableReaderForIndexJoin` missed it and how do other operators check it.

**Review by tiancaiamao — APPROVED** [review]
I'm afraid this is not the only place that missing the partition selection.
buildIndexReaderForIndexJoin / buildIndexLookupReaderForIndexJoin may have similiar issues...

**Comment by tiancaiamao** [comment]
> I think the better way is to go through the code to check why buildTableReaderForIndexJoin missed it and how do other operators check it.

The code is too complex, that's the real root cause for this kind of bug.
As you can see, the code sketch is ..

```
if v.IsCommonHandle() {
    if len(lookUpContents) > 0 && keyColumnsIncludeAllPartitionColumns(lookUpContents[0].keyCols, pe) {
        // buggy branch
    } else {
        partitionPruning()
    }
} else {
   if len(lookUpContents) > 0 && keyColumnsIncludeAllPartitionColumns(lookUpContents[0].keyCols, pe) {
        // buggy branch
   } else {
        partitionPruning()
   }
}
```

**Comment by mjonss** [comment]
@tiancaiamao @djshow832 I checked `buildIndexReaderForIndexJoin`, `buildIndexLookUpReaderForIndexJoin` and `buildTableReaderForIndexJoin` and they now all seems to call `partitionPruning` or `prunePartitionForInnerExecutor` before any if branches for dynamic partition prune mode.

**Review by djshow832 — APPROVED** [review]

**Pull request merged**